machine/v1beta1/types_controlplanemachineset.go

```diff
@@ -0,0 +1,244 @@
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+type ControlPlaneMachineSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControlPlaneMachineSetSpec   `json:"spec,omitempty"`
+	Status ControlPlaneMachineSetStatus `json:"status,omitempty"`
+}
+
+type ControlPlaneMachineSetSpec struct {
+	// Replicas defines how many Control Plane Machines should be
+	// created by this ControlPlaneMachineSet.
+	// This field is immutable and cannot be changed after cluster
+	// installation.
+	// The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
+	// 3 and 5 are the only valid values for this field.
+	// +kubebuilder:validation:Enum:=3;5
+	// +kubebuilder:default:=3
+	// +kubebuilder:validation:Required
+	Replicas *int32 `json:"replicas"`
+
+	// Strategy defines how the ControlPlaneMachineSet will update
+	// Machines when it detects a change to the ProviderSpec.
+	// +kubebuilder:default:={type: RollingUpdate}
+	// +optional
+	Strategy ControlPlaneMachineSetStrategy `json:"strategy,omitempty"`
+
+	// FailureDomains is the list of failure domains (sometimes called
+	// availability zones) in which the ControlPlaneMachineSet should balance
+	// the Control Plane Machines.
+	// This will be merged into the ProviderSpec given in the template.
+	// This field is optional on platforms that do not require placement
+	// information, eg OpenStack.
+	// +optional
+	FailureDomains FailureDomains `json:"failureDomains,omitempty"`
+
+	// Label selector for Machines. Existing Machines selected by this
+	// selector will be the ones affected by this ControlPlaneMachineSet.
+	// It must match the template's labels.
+	// This field is considered immutable after creation of the resource.
+	// +kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector"`
+
+	// Template describes the Control Plane Machines that will be created
+	// by this ControlPlaneMachineSet.
+	// +kubebuilder:validation:Required
+	Template ControlPlaneMachineSetTemplate `json:"template"`
+}
+
+type ControlPlaneMachineSetTemplate struct {
+	// ObjectMeta is the standard object metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// Labels are required to match the ControlPlaneMachineSet selector.
+	// +kubebuilder:validation:Required
+	ObjectMeta metav1.ObjectMeta `json:"metadata"`
+
+	// Spec contains the desired configuration of the Control Plane Machines.
+	// The ProviderSpec within contains platform specific details
+	// for creating the Control Plane Machines.
+	// The ProviderSe should be complete apart from the platform specific
+	// failure domain field. This will be overriden when the Machines
+	// are created based on the FailureDomains field.
+	// +kubebuilder:validation:Required
+	Spec MachineSpec `json:"spec"`
```

Comment:
!  not a reference, an actual MachineSpec?

Reply:
Yes, I hate this too! But, it matches our existing APIs (compare this CRD with the MachineSet CRD, they are very very similar). If we had a chance to start our API again, I would love to have a separate CRD for the MachineSpec that is then referenced (as CAPI does), but I don't think we can easily change our existing API to do this and I know historically there were a few strong voices against the proliferation of CRDs within the MAPI project.

I think it would be weird/confusing to have this as a v1 with a reference when the rest of our Machine API just inlines the specs